test/jdk/javax/net/ssl/TLSCommon/TLSWithEdDSA.java

```diff
@@ -0,0 +1,734 @@
+     */
+    @Override
+    protected void configureServerSocket(SSLServerSocket socket) {
+        String pVal;
+        try {
+            socket.setReuseAddress(true);
+            if ((pVal = serverParameters.get(ParamType.PROTOS)) != null) {
+                socket.setEnabledProtocols(pVal.split(":"));
+            }
+
+            if ((pVal = serverParameters.get(ParamType.CIPHERS)) != null) {
+                socket.setEnabledCipherSuites(pVal.split(":"));
+            }
+
+            pVal = serverParameters.get(ParamType.CLIAUTH);
+            socket.setWantClientAuth("WANT".equalsIgnoreCase(pVal));
+            socket.setNeedClientAuth("NEED".equalsIgnoreCase(pVal));
+        } catch (SocketException se) {
+            throw new RuntimeException(se);
+        }
+    }
+
+
+    @Override
+    protected void runServerApplication(SSLSocket socket) throws Exception {
+        InputStream sslIS = socket.getInputStream();
+        OutputStream sslOS = socket.getOutputStream();
+
+        sslIS.read();
+        sslOS.write(85);
+        sslOS.flush();
+
+        if (serverChecker != null) {
+            serverChecker.check(socket);
+        }
+    }
+
+    @Override
+    protected void runClientApplication(SSLSocket socket) throws Exception {
+        InputStream sslIS = socket.getInputStream();
+        OutputStream sslOS = socket.getOutputStream();
+
+        sslOS.write(280);
+        sslOS.flush();
+        sslIS.read();
+
+        if (clientChecker != null) {
+            clientChecker.check(socket);
+        }
+    }
+
+    public static void main(String[] args) throws Exception {
+        if (DEBUG) {
+            System.setProperty("javax.net.debug", "ssl:handshake");
+        }
+        certFac = CertificateFactory.getInstance("X.509");
+        String testFormat;
+
+        System.out.println("===== Test KeyManager alias retrieval =====");
+        keyManagerTests(DEF_ALL_EE, "EdDSA",
+                new String[] {"ee_ed25519", "ee_ed448"});
+
+        testFormat =
+                "===== Basic Ed25519 Server-side Authentication: %s =====\n";
+        serverParameters.put(ParamType.KSENTRIES, "EE_ED25519:EE_RSA_2048");
+        runtest(testFormat, isPeerEd25519, null, null, null);
+
+        testFormat =
+                "===== Basic Ed448 Server-side Authentication: %s =====\n";
+        serverParameters.put(ParamType.KSENTRIES, "EE_ED448:EE_RSA_2048");
+        runtest(testFormat, isPeerEd448, null, null, null);
+
+        testFormat = "===== EC favored over EdDSA by default: %s =====\n";
+        serverParameters.put(ParamType.KSENTRIES,
+                "EE_ED25519:EE_ECDSA_SECP521R1");
+        runtest(testFormat, isPeerP521, null, null, null);
+
+        testFormat = "===== Override EC favoring by alias: %s =====\n";
+        serverParameters.put(ParamType.CERTALIAS, "EE_ED25519");
+        runtest(testFormat, isPeerEd25519, null, null, null);
+        serverParameters.remove(ParamType.CERTALIAS);
```

Comment:
I did not get the idea here.  Is there a special case in practice that use a similar key manger like the AliasKeyManager?

Reply:
Right now, for TLS 1.0/1.1 EC certificates will be favored over EdDSA certificates in keystores that have valid certificates with both kinds of keys.  There's nothing we can do about that because 1.0/1.1 has no signaling mechanism to indicate signature preference like 1.2+ has.  Given that, I was thinking of ways to get around that restriction and one case I thought of was the Tomcat connector, which specifies a certificate to use by alias.  I wanted to make sure that we could still do that for 1.0/1.1 and it wouldn't break so I cooked up this simple KeyManager and ran a basic connection, expecting to see the cert specified by the alias.

Reply:
Got it.  Thanks!